### DISCUSSION TITLE
Cannot create shared_ptr with custom traits value

### DISCUSSION THREAD

**Discussion by golxzn** [taocpp/json#123] [Q&A] [answered]
Hi 👋

I've implemented my own trait like that:
```C++
template<class T>
struct CustomTraits : public tao::json::traits<T> {
	TAO_JSON_DEFAULT_KEY("custom");
};

template<> struct CustomTraits<SDL_FPoint> {
	TAO_JSON_DEFAULT_KEY("point");

	template<template<class...> class Traits>
	static void assign(tao::json::basic_value<Traits> &value, const SDL_FPoint &point);

	template<template<class...> class Traits>
	static SDL_FPoint as(const tao::json::basic_value<Traits> &value);

	template<template<class...> class Traits>
	static void to(const tao::json::basic_value<Traits> &value, SDL_FPoint &point);

	template<template<class...> class Traits>
	static bool equal(const tao::json::basic_value<Traits> &value, const SDL_FPoint &point) noexcept;

	template<template<class...> class Traits>
	static bool less_than(const tao::json::basic_value<Traits> &value, const SDL_FPoint &point) noexcept;

	template<template<class...> class Traits>
	static bool greater_than(const tao::json::basic_value<Traits> &value, const SDL_FPoint &point) noexcept;

	template<template<class...> class Traits, class Consumer>
	static void produce(Consumer &c, const SDL_FPoint &point);
};

template<class T>
using custom_trait = CustomTraits<T>;
using custom_value = tao::json::basic_value<CustomTraits>;
```
PS: (I hided the implementations to make this question shorter)

So, I'm trying to read the file and initialize my `custom_value` as a shared pointer:

```c++
std::shared_ptr<custom_value> config{};
if (fs::ifstream file{ origin, std::ios::binary | std::ios::ate }; file) {
	const auto size{ file.tellg() };
	file.seekg(std::ios::beg);
	std::string content(size, '\0');
	file.read(&content[0], size);
	config = std::make_shared<custom_value>(
		tao::json::basic_from_string<custom_traits>(std::move(content))
	);
}
```

But I get a lot of template errors. They are about `error C2672: 'std::construct_at': no matching overloaded function found` which means that the object custom_value doesn't have move constructor, but it has.

Before this error, I'd gotten the error inside `tao::json::internal::pair` that it cannot find an implementation of `Traits< std::decay_t< U > >::template default_key< Traits >::as_string()`, so I added TAO_JSON_DEFAULT_KEY into my traits classes.

Please, could you help me to initialize my custom `tao::json::basic_value<CustomTraits>` for the shared pointer for managing resources better.

**Comment by golxzn** [comment]
Oh. And `tao::json::basic_from_file` and `tao::json::basic_from_stream` don't work too.

**Comment by ColinH** [comment, marked as answer]
First, this problem has nothing to do with `std::shared_ptr`. Second, it works for me when I use `tao::json::basic_from_string<CustomTraits>` instead of `tao::json::basic_from_string<custom_trait>`.

The type alias `custom_trait` doesn't add anything, and I'm not really surprised that your example doesn't work given that templated type aliases aren't quite a replacement for class templates like normal type aliases are for classes/types, though I would have expected some different error.

And third, `basic_from_file` and `basic_from_stream` also work when using `CustomTraits` instead of `custom_trait`. Does this small change work for you, too?

**Reply by golxzn** [reply]
Yeah, sorry. I realized that the `tao::json::basic_from_string` template doesn't work with such type aliases
Thank you
